qa/integration-tests/src/test/java/io/camunda/it/client/FlowNodeInstanceQueryTest.java

```diff
@@ -0,0 +1,466 @@
+
+    // then
+    assertThat(result.items().size()).isEqualTo(4);
+    assertThat(result.items().getFirst().getState()).isEqualTo(state);
+  }
+
+  @Test
+  void shouldSortByState() {
+    // when
+    final var resultAsc =
+        zeebeClient.newFlownodeInstanceQuery().sort(s -> s.state().asc()).send().join();
+    final var resultDesc =
+        zeebeClient.newFlownodeInstanceQuery().sort(s -> s.state().desc()).send().join();
+
+    assertSorted(resultAsc, resultDesc, FlowNodeInstance::getState);
+  }
+
+  @Test
+  void shouldQueryByIncident() {
+    // given
+    final var hasIncident = flowNodeInstanceWithIncident.getIncident();
+    // when
+    final var result =
+        zeebeClient
+            .newFlownodeInstanceQuery()
+            .filter(f -> f.hasIncident(hasIncident))
+            .send()
+            .join();
+
+    // then
+    assertThat(result.items().size()).isEqualTo(7);
+    assertThat(result.items().getFirst().getIncident()).isEqualTo(hasIncident);
+  }
+
+  @Test
+  void shouldQueryByType() {
+    // given
+    final var type = flowNodeInstance.getType();
+    // when
+    final var result =
+        zeebeClient.newFlownodeInstanceQuery().filter(f -> f.type(type)).send().join();
+
+    // then
+    assertThat(result.items().size()).isEqualTo(3);
+    assertThat(result.items().getFirst().getType()).isEqualTo(type);
+    assertThat(result.items().get(1).getType()).isEqualTo(type);
+    assertThat(result.items().get(2).getType()).isEqualTo(type);
+  }
+
+  @Test
+  void shouldSortByType() {
+    // when
+    final var resultAsc =
+        zeebeClient.newFlownodeInstanceQuery().sort(s -> s.type().asc()).send().join();
+    final var resultDesc =
+        zeebeClient.newFlownodeInstanceQuery().sort(s -> s.type().desc()).send().join();
+
+    assertSorted(resultAsc, resultDesc, FlowNodeInstance::getType);
+  }
+
+  @Test
+  void shouldQueryByTreePath() {
+    // given
+    final var treePath = flowNodeInstance.getTreePath();
+    // when
+    final var result =
+        zeebeClient.newFlownodeInstanceQuery().filter(f -> f.treePath(treePath)).send().join();
+
+    // then
+    assertThat(result.items().size()).isEqualTo(1);
+    assertThat(result.items().getFirst().getTreePath()).isEqualTo(treePath);
+  }
+
+  @Test
+  void shouldQueryByTenantId() {
+    // given
+    final var tenantId = flowNodeInstance.getTenantId();
+    // when
+    final var result =
+        zeebeClient.newFlownodeInstanceQuery().filter(f -> f.tenantId(tenantId)).send().join();
+
+    // then
+    assertThat(result.items().size()).isEqualTo(7);
+    assertThat(result.items()).allMatch(f -> f.getTenantId().equals(tenantId));
+  }
+
+  @Test
+  public void shouldValidatePagination() {
+    final var result = zeebeClient.newFlownodeInstanceQuery().page(p -> p.limit(2)).send().join();
+    assertThat(result.items().size()).isEqualTo(2);
+    final var key = result.items().getFirst().getFlowNodeInstanceKey();
```

Comment:
❓ Shouldn't we have something like `firstSortValues` and `lastSortValues` that can be used to search before or after?

Reply:
Good point! Thanks.